File: presentation/src/main/java/com/smb/presentation/home/HomePresenter.java
package com.smb.presentation.home;

import android.content.Context;
import android.util.Log;

import com.arellomobile.mvp.InjectViewState;
import com.arellomobile.mvp.MvpPresenter;
import com.smb.R;
import com.smb.core.interactors.HomeDataUseCase;
import com.smb.core.models.HomePreviews;
import com.smb.core.models.Show;
import com.smb.core.models.holders.HomeScreenDataHolder;
import com.smb.core.models.util.Size;
import com.smb.core.repositories.ShowRepository;
import com.smb.di.DependencyContainer;
import com.smb.presentation.DefaultSize;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.functions.Consumer;
import io.reactivex.observers.DisposableSingleObserver;

@InjectViewState
public class HomePresenter extends MvpPresenter<HomeView> {
    @Inject
    HomeDataUseCase homeDataUseCase;
    @Inject
    ShowRepository showRepository;
    @Inject
    Context appContext;

    public HomePresenter() {
        super();
        DependencyContainer.getAppComponent().inject(this);
    }

    public void getData() {
        int bannerWidth = (int) appContext.getResources().getDimension(R.dimen.banner_width);
        int bannerHeight = (int) appContext.getResources().getDimension(R.dimen.banner_height);

        Size defaultSize = new DefaultSize(appContext);
        Size bannerSize = new Size(bannerWidth, bannerHeight);
        HomePreviews previews = new HomePreviews(defaultSize, bannerSize);
        homeDataUseCase.execute(previews, new DisposableSingleObserver<HomeScreenDataHolder>() {

            @Override
            public void onSuccess(HomeScreenDataHolder data) {
                getViewState().onDataFetched(data);
            }

            @Override
            public void onError(Throwable e) {

            }
        });

    }

    public void loadTest() {
        showRepository.getShows(new DefaultSize(appContext)).subscribe(new Consumer<List<Show>>() {
            @Override
            public void accept(List<Show> shows) throws Exception {
                Log.e("---", "accept: succcess");
            }
        }, new Consumer<Throwable>() {
            @Override
            public void accept(Throwable throwable) throws Exception {
                Log.e("---", "accept: error");
            }
        });
    }
}

File: presentation/src/main/java/com/smb/presentation/home/HomePresenter.kt
package com.smb.presentation.home

import android.content.Context
import com.arellomobile.mvp.InjectViewState
import com.arellomobile.mvp.MvpPresenter
import com.google.gson.Gson
import com.smb.data.repositories.api.LocalUserRepository
import com.smb.data.repositories.tv.TestShowsRepository
import com.smb.di.DependencyContainer
import com.smb.presentation.player.PlayerActivity
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject

@InjectViewState
class HomePresenter : MvpPresenter<HomeView>() {
    @Inject
    lateinit var localUserRepository: LocalUserRepository
    @Inject
    lateinit var testShowsRepository: TestShowsRepository

    override fun attachView(view: HomeView?) {
        super.attachView(view)
        DependencyContainer.getAppComponent().inject(this)
    }

    fun getUser(): String {
        return Gson().toJson(localUserRepository.get())
    }

    fun loadVideo(context: Context) {
        viewState.showProgress()
        testShowsRepository.randomHlsLink
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnTerminate { viewState.hideProgress() }
                .map { link -> PlayerActivity.getIntent(context, link) }
                .subscribe { intent -> viewState.startNewActivity(intent) }
    }
}
